In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [2]:
df = pd.read_csv("SampleSuperstore.csv")

In [3]:
print(df)

           Ship Mode    Segment        Country             City       State  \
0       Second Class   Consumer  United States        Henderson    Kentucky   
1       Second Class   Consumer  United States        Henderson    Kentucky   
2       Second Class  Corporate  United States      Los Angeles  California   
3     Standard Class   Consumer  United States  Fort Lauderdale     Florida   
4     Standard Class   Consumer  United States  Fort Lauderdale     Florida   
...              ...        ...            ...              ...         ...   
9989    Second Class   Consumer  United States            Miami     Florida   
9990  Standard Class   Consumer  United States       Costa Mesa  California   
9991  Standard Class   Consumer  United States       Costa Mesa  California   
9992  Standard Class   Consumer  United States       Costa Mesa  California   
9993    Second Class   Consumer  United States      Westminster  California   

      Postal Code Region         Category Sub-Categ

In [8]:
df.head()

,Ship Mode,Segment,Country,City,State,Postal Code,Region,Category,Sub-Category,Sales,Quantity,Discount,Profit
0,Second Class,Consumer,United States,Henderson,Kentucky,42420,South,Furniture,Bookcases,261.9600,2,0.00,41.9136
1,Second Class,Consumer,United States,Henderson,Kentucky,42420,South,Furniture,Chairs,731.9400,3,0.00,219.5820
2,Second Class,Corporate,United States,Los Angeles,California,90036,West,Office Supplies,Labels,14.6200,2,0.00,6.8714
3,Standard Class,Consumer,United States,Fort Lauderdale,Florida,33311,South,Furniture,Tables,957.5775,5,0.45,-383.0310
4,Standard Class,Consumer,United States,Fort Lauderdale,Florida,33311,South,Office Supplies,Storage,22.3680,2,0.20,2.5164


In [11]:
df.columns

Index(['Ship Mode', 'Segment', 'Country', 'City', 'State', 'Postal Code',
       'Region', 'Category', 'Sub-Category', 'Sales', 'Quantity', 'Discount',
       'Profit'],
      dtype='str')

In [13]:
df.shape

(9994, 13)

In [14]:
df.columns.unique()

Index(['Ship Mode', 'Segment', 'Country', 'City', 'State', 'Postal Code',
       'Region', 'Category', 'Sub-Category', 'Sales', 'Quantity', 'Discount',
       'Profit'],
      dtype='str')

In [16]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 9994 entries, 0 to 9993
Data columns (total 13 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Ship Mode     9994 non-null   str    
 1   Segment       9994 non-null   str    
 2   Country       9994 non-null   str    
 3   City          9994 non-null   str    
 4   State         9994 non-null   str    
 5   Postal Code   9994 non-null   int64  
 6   Region        9994 non-null   str    
 7   Category      9994 non-null   str    
 8   Sub-Category  9994 non-null   str    
 9   Sales         9994 non-null   float64
 10  Quantity      9994 non-null   int64  
 11  Discount      9994 non-null   float64
 12  Profit        9994 non-null   float64
dtypes: float64(3), int64(2), str(8)
memory usage: 1.7 MB


### Sorting the data in descending based on the profit

In [15]:
df.sort_values("Profit" , ascending =False )

,Ship Mode,Segment,Country,City,State,Postal Code,Region,Category,Sub-Category,Sales,Quantity,Discount,Profit
6826,Standard Class,Corporate,United States,Lafayette,Indiana,47905,Central,Technology,Copiers,17499.950,5,0.0,8399.9760
8153,First Class,Consumer,United States,Seattle,Washington,98115,West,Technology,Copiers,13999.960,4,0.0,6719.9808
4190,Standard Class,Consumer,United States,Newark,Delaware,19711,East,Technology,Copiers,10499.970,3,0.0,5039.9856
9039,Standard Class,Consumer,United States,Detroit,Michigan,48205,Central,Office Supplies,Binders,9892.740,13,0.0,4946.3700
4098,Standard Class,Consumer,United States,Minneapolis,Minnesota,55407,Central,Office Supplies,Binders,9449.950,5,0.0,4630.4755
...,...,...,...,...,...,...,...,...,...,...,...,...,...
4991,Standard Class,Corporate,United States,Chicago,Illinois,60653,Central,Office Supplies,Binders,1889.990,5,0.8,-2929.4845
3011,Standard Class,Home Office,United States,Louisville,Colorado,80027,West,Technology,Machines,2549.985,5,0.7,-3399.9800
9774,Standard Class,Consumer,United States,San Antonio,Texas,78207,Central,Office Supplies,Binders,2177.584,8,0.8,-3701.8928
683,Same Day,Corporate,United States,Burlington,North Carolina,27217,South,Technology,Machines,7999.980,4,0.5,-3839.9904


### Removing duplicate Values

In [17]:
df.duplicated().sum()

np.int64(17)

In [18]:
df = df.drop_duplicates()

In [21]:
df.duplicated().sum()

np.int64(0)

### Profit Margin

In [20]:
df["Profit Margin"] = (df["Profit"] / df["Sales"]) * 100

In [23]:
df.head()

,Ship Mode,Segment,Country,City,State,Postal Code,Region,Category,Sub-Category,Sales,Quantity,Discount,Profit,Profit Margin
0,Second Class,Consumer,United States,Henderson,Kentucky,42420,South,Furniture,Bookcases,261.9600,2,0.00,41.9136,16.00
1,Second Class,Consumer,United States,Henderson,Kentucky,42420,South,Furniture,Chairs,731.9400,3,0.00,219.5820,30.00
2,Second Class,Corporate,United States,Los Angeles,California,90036,West,Office Supplies,Labels,14.6200,2,0.00,6.8714,47.00
3,Standard Class,Consumer,United States,Fort Lauderdale,Florida,33311,South,Furniture,Tables,957.5775,5,0.45,-383.0310,-40.00
4,Standard Class,Consumer,United States,Fort Lauderdale,Florida,33311,South,Office Supplies,Storage,22.3680,2,0.20,2.5164,11.25


In [24]:
df.info()

<class 'pandas.DataFrame'>
Index: 9977 entries, 0 to 9993
Data columns (total 14 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Ship Mode      9977 non-null   str    
 1   Segment        9977 non-null   str    
 2   Country        9977 non-null   str    
 3   City           9977 non-null   str    
 4   State          9977 non-null   str    
 5   Postal Code    9977 non-null   int64  
 6   Region         9977 non-null   str    
 7   Category       9977 non-null   str    
 8   Sub-Category   9977 non-null   str    
 9   Sales          9977 non-null   float64
 10  Quantity       9977 non-null   int64  
 11  Discount       9977 non-null   float64
 12  Profit         9977 non-null   float64
 13  Profit Margin  9977 non-null   float64
dtypes: float64(4), int64(2), str(8)
memory usage: 1.9 MB


In [35]:
df[["Sales","Profit","Profit Margin"]].sort_values("Profit", ascending = False).head(10)

,Sales,Profit,Profit Margin
6826,17499.950,8399.9760,48.0
8153,13999.960,6719.9808,48.0
4190,10499.970,5039.9856,48.0
9039,9892.740,4946.3700,50.0
4098,9449.950,4630.4755,49.0
2623,11199.968,3919.9888,35.0
509,6354.950,3177.4750,50.0
8488,8749.950,2799.9840,32.0
7666,5399.910,2591.9568,48.0
6520,5443.960,2504.2216,46.0


In [32]:
df["Profit Margin"].describe()

count    9977.000000
mean       12.011354
std        46.663769
min      -275.000000
25%         7.500000
50%        27.000000
75%        36.250000
max        50.000000
Name: Profit Margin, dtype: float64

### Checking for zero values in sales and profit margin

In [36]:
(df["Sales"] <= 0).sum()

np.int64(0)

In [34]:
df["Profit Margin"].isna().sum()

np.int64(0)

### Overall KPIs

In [44]:
total_sales = df["Sales"].sum()
total_profit = df["Profit"].sum()
overall_margin=(total_profit / total_sales)*100

In [45]:
print("Total sales:", total_sales)
print("Total profit:", total_profit)
print("OveraLL Margin:",overall_margin)

Total sales: 2296195.5903
Total profit: 286241.42260000005
OveraLL Margin: 12.46589897695093


In [46]:
df[["Sales","Profit","Profit Margin","Discount","Quantity"]].describe()

,Sales,Profit,Profit Margin,Discount,Quantity
count,9977.000000,9977.00000,9977.000000,9977.000000,9977.000000
mean,230.148902,28.69013,12.011354,0.156278,3.790719
std,623.721409,234.45784,46.663769,0.206455,2.226657
min,0.444000,-6599.97800,-275.000000,0.000000,1.000000
25%,17.300000,1.72620,7.500000,0.000000,2.000000
50%,54.816000,8.67100,27.000000,0.200000,3.000000
75%,209.970000,29.37200,36.250000,0.200000,5.000000
max,22638.480000,8399.97600,50.000000,0.800000,14.000000


In [47]:
negative_profit = df[df["Profit"] < 0]

print("Negative-profit transactions:", len(negative_profit))
print("Total loss:", negative_profit["Profit"].sum())

Negative-profit transactions: 1869
Total loss: -156112.98990000002


### Where is the loss coming From?

In [50]:
category_analysis = (df.groupby("Category").agg(Sales=("Sales","sum"),
                                               Profit=("Profit","sum"),
                                               Quantity=("Quantity","sum")))
category_analysis["Profit Margin"] = (category_analysis["Profit"]/category_analysis["Sales"])*100
category_analysis

,Sales,Profit,Quantity,Profit Margin
Category,,,,
Furniture,741306.3133,18421.8137,8020,2.485047
Office Supplies,718735.2440,122364.6608,22861,17.024998
Technology,836154.0330,145454.9481,6939,17.395712


In [52]:
category_analysis.sort_values("Profit")


,Sales,Profit,Quantity,Profit Margin
Category,,,,
Furniture,741306.3133,18421.8137,8020,2.485047
Office Supplies,718735.2440,122364.6608,22861,17.024998
Technology,836154.0330,145454.9481,6939,17.395712


In [53]:
category_loss=(df[df["Profit"] < 0].groupby("Category").agg(Loss_Transactions=("Profit","count"),
                                                            Total_loss=("Profit","sum")).sort_values("Total_loss"))

In [55]:
category_loss

,Loss_Transactions,Total_loss
Category,,
Furniture,713,-60924.0502
Office Supplies,885,-56609.0215
Technology,271,-38579.9182


In [60]:
subcategory_analysis =(df.groupby("Sub-Category").agg(Sales=("Sales","sum"),
                                                      Profit=("Profit","sum"),
                                                      Quantity=("Quantity","sum")))
subcategory_analysis["Profit Margin"] = (subcategory_analysis["Profit"]/subcategory_analysis["Sales"]) * 100
subcategory_analysis

,Sales,Profit,Quantity,Profit Margin
Sub-Category,,,,
Accessories,167380.3180,41936.6357,2976,25.054700
Appliances,107532.1610,18138.0054,1729,16.867517
Art,27107.0320,6524.6118,2996,24.069813
Binders,203409.1690,30228.0003,5971,14.860687
Bookcases,114879.9963,-3472.5560,868,-3.022768
Chairs,327777.7610,26567.1278,2351,8.105226
Copiers,149528.0300,55617.8249,234,37.195585
Envelopes,16476.4020,6964.1767,906,42.267582
Fasteners,3024.2800,949.5182,914,31.396504


In [61]:
subcategory_loss = (
    df[df["Profit"] < 0]
    .groupby("Sub-Category")
    .agg(
        Loss_Transactions=("Profit", "count"),
        Total_Loss=("Profit", "sum")
    )
    .sort_values("Total_Loss")
)

subcategory_loss

,Loss_Transactions,Total_Loss
Sub-Category,,
Binders,612,-38504.2594
Tables,203,-32412.1483
Machines,44,-30118.6682
Bookcases,109,-12152.2060
Chairs,234,-9868.7825
Appliances,67,-8629.6412
Phones,136,-7530.6235
Furnishings,167,-6490.9134
Storage,161,-6426.3038


### Discount Analysis

In [62]:
df["Discount Band"] = pd.cut(
    df["Discount"],
    bins=[-0.01, 0.10, 0.20, 0.30, 0.40, 1],
    labels=["0-10%", "10-20%", "20-30%", "30-40%", "40%+"]
)

discount_analysis = (
    df.groupby("Discount Band", observed=True)
      .agg(
          Revenue=("Sales", "sum"),
          Profit=("Profit", "sum"),
          Records=("Profit", "count")
      )
)

discount_analysis["Profit Margin"] = (
    discount_analysis["Profit"] /
    discount_analysis["Revenue"]
) * 100

discount_analysis

,Revenue,Profit,Records,Profit Margin
Discount Band,,,,
0-10%,1.141647e+06,329873.5843,4881,28.894537
10-20%,7.920635e+05,91725.5985,3705,11.580587
20-30%,1.029453e+05,-10357.2186,226,-10.060897
30-40%,1.309112e+05,-25448.1881,233,-19.439269
40%+,1.286287e+05,-99552.3535,932,-77.395140


In [63]:
high_discount_loss = (
    df[
        (df["Discount"] > 0.20) &
        (df["Profit"] < 0)
    ]
    .groupby("Sub-Category")
    .agg(
        Loss_Records=("Profit", "count"),
        Total_Loss=("Profit", "sum"),
        Revenue=("Sales", "sum")
    )
)

high_discount_loss["Loss Margin"] = (
    high_discount_loss["Total_Loss"] /
    high_discount_loss["Revenue"]
) * 100

high_discount_loss = high_discount_loss.sort_values("Total_Loss")

high_discount_loss

,Loss_Records,Total_Loss,Revenue,Loss Margin
Sub-Category,,,,
Binders,612,-38504.2594,36137.0490,-106.550647
Tables,172,-30761.1238,88219.0420,-34.869030
Machines,44,-30118.6682,72456.2530,-41.568073
Bookcases,69,-11097.7614,28345.2548,-39.152096
Appliances,67,-8629.6412,3382.5340,-255.123561
Chairs,147,-6725.0579,63167.4960,-10.646390
Phones,97,-6715.7800,31979.1600,-21.000489
Furnishings,138,-5944.6552,6644.7000,-89.464614


In [64]:
print("Negative Profit:",
      (df["Profit"] < 0).sum())

print("Discount > 20%:",
      (df["Discount"] > 0.20).sum())

print("Discount > 40%:",
      (df["Discount"] > 0.40).sum())

print("Negative Margin:",
      (df["Profit Margin"] < 0).sum())

Negative Profit: 1869
Discount > 20%: 1391
Discount > 40%: 932
Negative Margin: 1869


In [65]:
df["Leak Score"] = 0

In [66]:
# Negative profitability
df.loc[df["Profit"] < 0, "Leak Score"] += 50

In [67]:
# High discount
df.loc[df["Discount"] > 0.20, "Leak Score"] += 20

In [68]:
# Very high discount
df.loc[df["Discount"] > 0.40, "Leak Score"] += 20

In [69]:
# Severe negative margin
df.loc[df["Profit Margin"] < -20, "Leak Score"] += 10

In [70]:
df[
    df["Leak Score"] >= 70
].sort_values(
    "Leak Score",
    ascending=False
).head(20)

,Ship Mode,Segment,Country,City,State,Postal Code,Region,Category,Sub-Category,Sales,Quantity,Discount,Profit,Profit Margin,Discount Band,Leak Score
9921,Standard Class,Home Office,United States,Akron,Ohio,44312,East,Office Supplies,Binders,5.742,3,0.7,-4.5936,-80.000000,40%+,100
9795,Standard Class,Corporate,United States,Chicago,Illinois,60610,Central,Office Supplies,Binders,3.798,3,0.8,-5.8869,-155.000000,40%+,100
9794,Second Class,Consumer,United States,Houston,Texas,77070,Central,Office Supplies,Binders,18.240,3,0.8,-31.0080,-170.000000,40%+,100
9779,Standard Class,Consumer,United States,San Antonio,Texas,78207,Central,Office Supplies,Appliances,4.836,2,0.8,-12.0900,-250.000000,40%+,100
9778,Standard Class,Consumer,United States,San Antonio,Texas,78207,Central,Office Supplies,Binders,8.880,5,0.8,-13.3200,-150.000000,40%+,100
9776,Standard Class,Consumer,United States,San Antonio,Texas,78207,Central,Office Supplies,Binders,16.784,4,0.8,-26.8544,-160.000000,40%+,100
9775,Standard Class,Consumer,United States,San Antonio,Texas,78207,Central,Furniture,Furnishings,17.496,3,0.6,-10.0602,-57.500000,40%+,100
9774,Standard Class,Consumer,United States,San Antonio,Texas,78207,Central,Office Supplies,Binders,2177.584,8,0.8,-3701.8928,-170.000000,40%+,100
9747,Standard Class,Consumer,United States,Chicago,Illinois,60653,Central,Furniture,Furnishings,10.984,2,0.6,-7.9634,-72.500000,40%+,100
9711,Standard Class,Consumer,United States,Philadelphia,Pennsylvania,19143,East,Office Supplies,Binders,13.194,2,0.7,-8.7960,-66.666667,40%+,100


In [71]:
df["Leak Score"].value_counts().sort_index()

Leak Score
0      8063
20       45
50      393
60      130
70      231
80      183
90        7
100     925
Name: count, dtype: int64

### Financial Impact by score

In [73]:
score_analysis = (
    df.groupby("Leak Score").agg(Records=("Profit", "count"),
                                 Revenue=("Sales", "sum"),
                                 Profit=("Profit", "sum")))

score_analysis["Profit Margin"] = (score_analysis["Profit"] / score_analysis["Revenue"]) * 100

score_analysis

,Records,Revenue,Profit,Profit Margin
Leak Score,,,,
0,8063,1.795620e+06,439215.2256,24.460371
20,45,3.215373e+04,3139.1869,9.763058
50,393,1.124506e+05,-11741.1250,-10.441141
60,130,2.564014e+04,-5874.9178,-22.912974
70,231,1.225580e+05,-13869.8231,-11.316943
80,183,7.914476e+04,-25074.7705,-31.682161
90,7,2.509564e+04,-1894.1690,-7.547801
100,925,1.035330e+05,-97658.1845,-94.325615


### Actual Profit leak

In [74]:
df["Profit Leak"] = np.where(
    (df["Leak Score"] >= 50) & (df["Profit"] < 0),
    "Yes",
    "No"
)

df["Profit Leak"].value_counts()

Profit Leak
No     8108
Yes    1869
Name: count, dtype: int64

In [75]:
leak_summary = (
    df.groupby("Profit Leak")
      .agg(
          Records=("Profit", "count"),
          Revenue=("Sales", "sum"),
          Profit=("Profit", "sum")
      )
)

leak_summary["Profit Margin"] = (
    leak_summary["Profit"] /
    leak_summary["Revenue"]
) * 100

leak_summary

,Records,Revenue,Profit,Profit Margin
Profit Leak,,,,
No,8108,1.827773e+06,442354.4125,24.201820
Yes,1869,4.684222e+05,-156112.9899,-33.327409


### Classify Leaks

In [77]:
def classify_leak(row):
    if row["Profit"] >= 0:
        return "No Leak"
    if row["Discount"] > 0.40:
        return "Extreme Discount"
    if row["Discount"] > 0.20:
        return "High Discount"
    if row["Profit Margin"] < -20:
        return "Severe Margin Loss"
    return "Low Margin Loss"
df["Leak Type"] = df.apply(classify_leak, axis=1)

In [78]:
leak_type_analysis = (
    df.groupby("Leak Type")
      .agg(
          Records=("Profit", "count"),
          Revenue=("Sales", "sum"),
          Profit=("Profit", "sum")))

leak_type_analysis["Profit Margin"] = (
    leak_type_analysis["Profit"] /
    leak_type_analysis["Revenue"]) * 100

leak_type_analysis.sort_values("Profit")

,Records,Revenue,Profit,Profit Margin
Leak Type,,,,
Extreme Discount,932,1.286287e+05,-99552.3535,-77.395140
High Discount,414,2.017028e+05,-38944.5936,-19.307909
Low Margin Loss,393,1.124506e+05,-11741.1250,-10.441141
Severe Margin Loss,130,2.564014e+04,-5874.9178,-22.912974
No Leak,8108,1.827773e+06,442354.4125,24.201820


### Where Should a Business Investigate First 
#### we will combine:
- Number of Loss Records
- Total Financial loss
- Loss Margin

In [79]:
leak_impact = (
    df[df["Profit"] < 0]
    .groupby("Sub-Category")
    .agg(
        Loss_Records=("Profit", "count"),
        Total_Loss=("Profit", "sum"),
        Revenue=("Sales", "sum")
    )
)

leak_impact["Loss_Per_Record"] = (
    abs(leak_impact["Total_Loss"]) /
    leak_impact["Loss_Records"]
)

leak_impact["Loss_Margin"] = (
    leak_impact["Total_Loss"] /
    leak_impact["Revenue"]
) * 100

leak_impact.sort_values(
    "Total_Loss"
)

,Loss_Records,Total_Loss,Revenue,Loss_Per_Record,Loss_Margin
Sub-Category,,,,,
Binders,612,-38504.2594,36137.0490,62.915457,-106.550647
Tables,203,-32412.1483,104978.5460,159.665755,-30.875021
Machines,44,-30118.6682,72456.2530,684.515186,-41.568073
Bookcases,109,-12152.2060,48072.7408,111.488128,-25.278788
Chairs,234,-9868.7825,91707.0840,42.174284,-10.761200
Appliances,67,-8629.6412,3382.5340,128.800615,-255.123561
Phones,136,-7530.6235,35797.8400,55.372232,-21.036530
Furnishings,167,-6490.9134,12845.8440,38.867745,-50.529287
Storage,161,-6426.3038,37869.0720,39.914930,-16.969795


## Build Profit Leak Score
- Total loss → overall financial impact
- Loss records → frequency
- Loss per record → severity

In [81]:
leak_impact["Loss_Share"] = (
    abs(leak_impact["Total_Loss"]) /
    abs(leak_impact["Total_Loss"]).sum()
) * 100

leak_impact

,Loss_Records,Total_Loss,Revenue,Loss_Per_Record,Loss_Margin,Loss_Share
Sub-Category,,,,,,
Accessories,91,-930.6265,10958.8000,10.226665,-8.492047,0.596124
Appliances,67,-8629.6412,3382.5340,128.800615,-255.123561,5.527818
Binders,612,-38504.2594,36137.0490,62.915457,-106.550647,24.664353
Bookcases,109,-12152.2060,48072.7408,111.488128,-25.278788,7.784238
Chairs,234,-9868.7825,91707.0840,42.174284,-10.761200,6.321564
Fasteners,12,-33.1952,149.2800,2.766267,-22.236870,0.021264
Furnishings,167,-6490.9134,12845.8440,38.867745,-50.529287,4.157830
Machines,44,-30118.6682,72456.2530,684.515186,-41.568073,19.292865
Phones,136,-7530.6235,35797.8400,55.372232,-21.036530,4.823829


### Step 6 — Create a ProfitGuard Priority Score

We'll normalize the three measures so they're comparable:

- Loss Share → financial impact
- Loss Records → frequency
- Loss Per Record → severity

In [82]:
leak_impact["Impact_Score"] = (
    leak_impact["Loss_Share"] /
    leak_impact["Loss_Share"].max()
) * 100

leak_impact["Frequency_Score"] = (
    leak_impact["Loss_Records"] /
    leak_impact["Loss_Records"].max()
) * 100

leak_impact["Severity_Score"] = (
    leak_impact["Loss_Per_Record"] /
    leak_impact["Loss_Per_Record"].max()
) * 100

leak_impact

,Loss_Records,Total_Loss,Revenue,Loss_Per_Record,Loss_Margin,Loss_Share,Impact_Score,Frequency_Score,Severity_Score
Sub-Category,,,,,,,,,
Accessories,91,-930.6265,10958.8000,10.226665,-8.492047,0.596124,2.416944,14.869281,1.494001
Appliances,67,-8629.6412,3382.5340,128.800615,-255.123561,5.527818,22.412173,10.947712,18.816327
Binders,612,-38504.2594,36137.0490,62.915457,-106.550647,24.664353,100.000000,100.000000,9.191243
Bookcases,109,-12152.2060,48072.7408,111.488128,-25.278788,7.784238,31.560680,17.810458,16.287167
Chairs,234,-9868.7825,91707.0840,42.174284,-10.761200,6.321564,25.630366,38.235294,6.161190
Fasteners,12,-33.1952,149.2800,2.766267,-22.236870,0.021264,0.086212,1.960784,0.404121
Furnishings,167,-6490.9134,12845.8440,38.867745,-50.529287,4.157830,16.857650,27.287582,5.678142
Machines,44,-30118.6682,72456.2530,684.515186,-41.568073,19.292865,78.221653,7.189542,100.000000
Phones,136,-7530.6235,35797.8400,55.372232,-21.036530,4.823829,19.557897,22.222222,8.089263


### Create he Priority Score

In [83]:
leak_impact["Priority_Score"] = (
    0.50 * leak_impact["Impact_Score"] +
    0.30 * leak_impact["Frequency_Score"] +
    0.20 * leak_impact["Severity_Score"]
)

leak_impact = leak_impact.sort_values(
    "Priority_Score",
    ascending=False
)

leak_impact

,Loss_Records,Total_Loss,Revenue,Loss_Per_Record,Loss_Margin,Loss_Share,Impact_Score,Frequency_Score,Severity_Score,Priority_Score
Sub-Category,,,,,,,,,,
Binders,612,-38504.2594,36137.0490,62.915457,-106.550647,24.664353,100.000000,100.000000,9.191243,81.838249
Machines,44,-30118.6682,72456.2530,684.515186,-41.568073,19.292865,78.221653,7.189542,100.000000,61.267689
Tables,203,-32412.1483,104978.5460,159.665755,-30.875021,20.761980,84.178085,33.169935,23.325378,56.705099
Chairs,234,-9868.7825,91707.0840,42.174284,-10.761200,6.321564,25.630366,38.235294,6.161190,25.518009
Bookcases,109,-12152.2060,48072.7408,111.488128,-25.278788,7.784238,31.560680,17.810458,16.287167,24.380910
Appliances,67,-8629.6412,3382.5340,128.800615,-255.123561,5.527818,22.412173,10.947712,18.816327,18.253666
Phones,136,-7530.6235,35797.8400,55.372232,-21.036530,4.823829,19.557897,22.222222,8.089263,18.063468
Furnishings,167,-6490.9134,12845.8440,38.867745,-50.529287,4.157830,16.857650,27.287582,5.678142,17.750728
Storage,161,-6426.3038,37869.0720,39.914930,-16.969795,4.116444,16.689852,26.307190,5.831124,17.403308


### Three business patterns

- Binders = recurring problem

612 loss-making records and 24.66% of total loss makes this a repeated profitability issue.

Machines = high-severity problem

Only 44 loss records, but each averages about ₹684.52 loss. This is exactly why severity belongs in the model.

Tables = high-impact problem

₹32.4K of losses and 20.76% of total loss make it another major area to investigate.

So ProfitGuard can now distinguish:

 "Where are losses happening most often?"

from

 "Where are losses financially concentrated?"

from

"Where is each problematic transaction unusually severe?"

That's a meaningful improvement over simply filtering Profit < 0.

## Anomaly Detection

In [84]:
from sklearn.ensemble import IsolationForest

ml_features = df[
    ["Sales", "Quantity", "Discount", "Profit", "Profit Margin"]
].copy()

ml_features = ml_features.replace(
    [np.inf, -np.inf],
    np.nan
).dropna()

print("Rows available for ML:", len(ml_features))
ml_features.head()

Rows available for ML: 9977


,Sales,Quantity,Discount,Profit,Profit Margin
0,261.9600,2,0.00,41.9136,16.00
1,731.9400,3,0.00,219.5820,30.00
2,14.6200,2,0.00,6.8714,47.00
3,957.5775,5,0.45,-383.0310,-40.00
4,22.3680,2,0.20,2.5164,11.25


In [85]:
model = IsolationForest(
    contamination=0.05,
    random_state=42
)

model.fit(ml_features)

df.loc[ml_features.index, "Anomaly"] = model.predict(ml_features)

df["Anomaly"].value_counts()

Anomaly
 1.0    9478
-1.0     499
Name: count, dtype: int64

## anomaly analysis

In [87]:
anomaly_analysis = (
    df.groupby("Anomaly")
      .agg(
          Records=("Profit", "count"),
          Revenue=("Sales", "sum"),
          Profit=("Profit", "sum"),
          Loss_Records=("Profit", lambda x: (x < 0).sum())
      )
)

anomaly_analysis["Profit Margin"] = (
    anomaly_analysis["Profit"] /
    anomaly_analysis["Revenue"]
) * 100

anomaly_analysis["Loss Rate"] = (
    anomaly_analysis["Loss_Records"] /
    anomaly_analysis["Records"]
) * 100

anomaly_analysis

,Records,Revenue,Profit,Loss_Records,Profit Margin,Loss Rate
Anomaly,,,,,,
-1.0,499,7.191340e+05,46856.6637,328,6.515706,65.731463
1.0,9478,1.577062e+06,239384.7589,1541,15.179164,16.258704


### The important finding is the Loss Rate:

- Anomalies → 65.73%
- Normal → 16.26%

## ML-backed ProfitGuard signal

In [88]:
df["ML_Profit_Leak"] = np.where(
    (df["Anomaly"] == -1) &
    (df["Profit"] < 0),
    "Yes",
    "No"
)

ml_leaks = df[df["ML_Profit_Leak"] == "Yes"]

print("ML Profit Leak Candidates:", len(ml_leaks))
print("Total Loss:", ml_leaks["Profit"].sum())

ML Profit Leak Candidates: 328
Total Loss: -106937.1006


### Instead of our earlier definition:

Profit < 0 → Profit Leak

we now have:

Unusual transaction + negative profit → ML Profit Leak Candidate

<p> ProfitGuard uses unsupervised anomaly detection to identify unusual transactions that are disproportionately associated with negative profitability, creating a focused investigation queue rather than flagging every loss-making transaction.
</p>

## Give every ML leak an explanation

In [89]:
def explain_leak(row):
    reasons = []

    if row["Profit"] < 0:
        reasons.append("Negative Profit")

    if row["Discount"] > 0.40:
        reasons.append("Extreme Discount")
    elif row["Discount"] > 0.20:
        reasons.append("High Discount")

    if row["Profit Margin"] < -20:
        reasons.append("Severe Margin Loss")

    if row["Anomaly"] == -1:
        reasons.append("Unusual Transaction Pattern")

    return " + ".join(reasons)


df["Leak Reason"] = df.apply(explain_leak, axis=1)

ml_leaks = df[df["ML_Profit_Leak"] == "Yes"].copy()

ml_leaks[
    [
        "Sub-Category",
        "Sales",
        "Quantity",
        "Discount",
        "Profit",
        "Profit Margin",
        "Leak Reason"
    ]
].head(10)

,Sub-Category,Sales,Quantity,Discount,Profit,Profit Margin,Leak Reason
3,Tables,957.5775,5,0.45,-383.0310,-40.000000,Negative Profit + Extreme Discount + Severe Ma...
14,Appliances,68.8100,5,0.80,-123.8580,-180.000000,Negative Profit + Extreme Discount + Severe Ma...
27,Bookcases,3083.4300,7,0.50,-1665.0522,-54.000000,Negative Profit + Extreme Discount + Severe Ma...
118,Binders,157.7940,1,0.70,-115.7156,-73.333333,Negative Profit + Extreme Discount + Severe Ma...
125,Tables,617.7000,6,0.50,-407.6820,-66.000000,Negative Profit + Extreme Discount + Severe Ma...
165,Machines,8159.9520,8,0.40,-1359.9920,-16.666667,Negative Profit + High Discount + Unusual Tran...
167,Chairs,1740.0600,9,0.30,-24.8580,-1.428571,Negative Profit + High Discount + Unusual Tran...
169,Appliances,177.9800,5,0.80,-453.8490,-255.000000,Negative Profit + Extreme Discount + Severe Ma...
174,Appliances,52.4480,2,0.80,-131.1200,-250.000000,Negative Profit + Extreme Discount + Severe Ma...
176,Appliances,97.2640,4,0.80,-243.1600,-250.000000,Negative Profit + Extreme Discount + Severe Ma...


## final ProfitGuard flags

In [91]:
df["Negative_Profit"] = df["Profit"] < 0

df["High_Discount"] = df["Discount"] > 0.20

df["Severe_Margin"] = df["Profit Margin"] < -20

df["ML_Anomaly"] = df["Anomaly"] == -1

df["Leak_Candidate"] = (
    df["Negative_Profit"] &
    df["ML_Anomaly"]
)

df[
    [
        "Sub-Category",
        "Profit",
        "Discount",
        "Profit Margin",
        "Negative_Profit",
        "High_Discount",
        "Severe_Margin",
        "ML_Anomaly",
        "Leak_Candidate"
    ]
].head(10)

,Sub-Category,Profit,Discount,Profit Margin,Negative_Profit,High_Discount,Severe_Margin,ML_Anomaly,Leak_Candidate
0,Bookcases,41.9136,0.00,16.00,False,False,False,False,False
1,Chairs,219.5820,0.00,30.00,False,False,False,False,False
2,Labels,6.8714,0.00,47.00,False,False,False,False,False
3,Tables,-383.0310,0.45,-40.00,True,True,True,True,True
4,Storage,2.5164,0.20,11.25,False,False,False,False,False
5,Furnishings,14.1694,0.00,29.00,False,False,False,False,False
6,Art,1.9656,0.00,27.00,False,False,False,False,False
7,Phones,90.7152,0.20,10.00,False,False,False,False,False
8,Binders,5.7825,0.20,31.25,False,False,False,False,False
9,Appliances,34.4700,0.00,30.00,False,False,False,False,False


## Evidence Score 

In [92]:
df["Evidence_Count"] = (
    df["Negative_Profit"].astype(int) +
    df["High_Discount"].astype(int) +
    df["Severe_Margin"].astype(int) +
    df["ML_Anomaly"].astype(int)
)

df[
    [
        "Sub-Category",
        "Profit",
        "Discount",
        "Profit Margin",
        "Evidence_Count",
        "Leak_Candidate",
        "Leak Reason"
    ]
].sort_values(
    "Evidence_Count",
    ascending=False
).head(20)

,Sub-Category,Profit,Discount,Profit Margin,Evidence_Count,Leak_Candidate,Leak Reason
3084,Binders,-417.0936,0.8,-165.000000,4,True,Negative Profit + Extreme Discount + Severe Ma...
3968,Binders,-9.1648,0.8,-160.000000,4,True,Negative Profit + Extreme Discount + Severe Ma...
4186,Binders,-10.6547,0.8,-155.000000,4,True,Negative Profit + Extreme Discount + Severe Ma...
4980,Appliances,-11.7612,0.8,-270.000000,4,True,Negative Profit + Extreme Discount + Severe Ma...
2507,Appliances,-85.2384,0.8,-260.000000,4,True,Negative Profit + Extreme Discount + Severe Ma...
7554,Bookcases,-216.9783,0.5,-62.000000,4,True,Negative Profit + Extreme Discount + Severe Ma...
4187,Binders,-3.2934,0.8,-165.000000,4,True,Negative Profit + Extreme Discount + Severe Ma...
2628,Machines,-935.9595,0.7,-143.333333,4,True,Negative Profit + Extreme Discount + Severe Ma...
4614,Appliances,-166.3935,0.8,-265.000000,4,True,Negative Profit + Extreme Discount + Severe Ma...
5048,Binders,-13.8278,0.8,-170.000000,4,True,Negative Profit + Extreme Discount + Severe Ma...


<p>Why this is useful

A transaction with:

- Evidence Count = 4

<u>has all four warning signals.</u>

A transaction with:

- Evidence Count = 2

<u>has two independent signals.</u>

And:

- Evidence Count = 0

<u>has none.</u>

This gives us an explainable evidence-strength measure, rather than pretending the number is a statistically calibrated probability.</p>

### Validate Evidence  Count Financially

What we're checking

We want to see whether higher evidence counts generally correspond to:

higher loss rates
worse profit margins
greater total losses

In [94]:
evidence_analysis = (
    df.groupby("Evidence_Count")
      .agg(
          Records=("Profit", "count"),
          Revenue=("Sales", "sum"),
          Profit=("Profit", "sum"),
          Loss_Records=("Profit", lambda x: (x < 0).sum())
      )
)

evidence_analysis["Profit Margin"] = (
    evidence_analysis["Profit"] /
    evidence_analysis["Revenue"]
) * 100

evidence_analysis["Loss Rate"] = (
    evidence_analysis["Loss_Records"] /
    evidence_analysis["Records"]
) * 100

evidence_analysis

,Records,Revenue,Profit,Loss_Records,Profit Margin,Loss Rate
Evidence_Count,,,,,,
0,7896,1.287995e+06,286854.2429,0,22.271384,0.000000
1,594,6.238759e+05,144637.1304,386,23.183637,64.983165
2,362,1.381821e+05,-13986.9692,358,-10.122128,98.895028
3,821,1.264522e+05,-35166.7748,821,-27.810335,100.000000
4,304,1.196907e+05,-96096.2067,304,-80.287125,100.000000


#### Create the final ProfitGuard leak dataset

In [96]:
profit_leaks = df[df["Leak_Candidate"]].copy()

print("Total Profit Leak Candidates:", len(profit_leaks))
print("Total Revenue:", profit_leaks["Sales"].sum())
print("Total Loss:", profit_leaks["Profit"].sum())
print(
    "Loss as % of Total Observed Loss:",
    abs(profit_leaks["Profit"].sum()) /
    abs(df[df["Profit"] < 0]["Profit"].sum()) * 100
)

Total Profit Leak Candidates: 328
Total Revenue: 199226.14279999997
Total Loss: -106937.1006
Loss as % of Total Observed Loss: 68.49980944474883


### Sub-Category Priority Intelligence

In [98]:
priority_map = leak_impact["Priority_Score"].to_dict()

df["Subcategory_Priority"] = (
    df["Sub-Category"]
    .map(priority_map)
    .fillna(0)
)

df[
    [
        "Sub-Category",
        "Profit",
        "Leak_Candidate",
        "Subcategory_Priority"
    ]
].head(20)

,Sub-Category,Profit,Leak_Candidate,Subcategory_Priority
0,Bookcases,41.9136,False,24.380910
1,Chairs,219.5820,False,25.518009
2,Labels,6.8714,False,0.000000
3,Tables,-383.0310,True,56.705099
4,Storage,2.5164,False,17.403308
5,Furnishings,14.1694,False,17.750728
6,Art,1.9656,False,0.000000
7,Phones,90.7152,False,18.063468
8,Binders,5.7825,False,81.838249
9,Appliances,34.4700,False,18.253666


### Final PowerBI dataset

In [99]:
profit_leaks = df[df["Leak_Candidate"]].copy()

profit_leaks = profit_leaks[
    [
        "Ship Mode",
        "Segment",
        "Country",
        "City",
        "State",
        "Postal Code",
        "Region",
        "Category",
        "Sub-Category",
        "Sales",
        "Quantity",
        "Discount",
        "Profit",
        "Profit Margin",
        "Discount Band",
        "Anomaly",
        "Negative_Profit",
        "High_Discount",
        "Severe_Margin",
        "ML_Anomaly",
        "Evidence_Count",
        "Leak_Candidate",
        "Leak Type",
        "Leak Reason",
        "Subcategory_Priority"
    ]
]

print("Final rows:", len(profit_leaks))
print("Final columns:", len(profit_leaks.columns))

profit_leaks.head()

Final rows: 328
Final columns: 25


,Ship Mode,Segment,Country,City,State,Postal Code,Region,Category,Sub-Category,Sales,...,Anomaly,Negative_Profit,High_Discount,Severe_Margin,ML_Anomaly,Evidence_Count,Leak_Candidate,Leak Type,Leak Reason,Subcategory_Priority
3,Standard Class,Consumer,United States,Fort Lauderdale,Florida,33311,South,Furniture,Tables,957.5775,...,-1.0,True,True,True,True,4,True,Extreme Discount,Negative Profit + Extreme Discount + Severe Ma...,56.705099
14,Standard Class,Home Office,United States,Fort Worth,Texas,76106,Central,Office Supplies,Appliances,68.8100,...,-1.0,True,True,True,True,4,True,Extreme Discount,Negative Profit + Extreme Discount + Severe Ma...,18.253666
27,Standard Class,Consumer,United States,Philadelphia,Pennsylvania,19140,East,Furniture,Bookcases,3083.4300,...,-1.0,True,True,True,True,4,True,Extreme Discount,Negative Profit + Extreme Discount + Severe Ma...,24.380910
118,Standard Class,Corporate,United States,Bristol,Tennessee,37620,South,Office Supplies,Binders,157.7940,...,-1.0,True,True,True,True,4,True,Extreme Discount,Negative Profit + Extreme Discount + Severe Ma...,81.838249
125,Standard Class,Consumer,United States,Bloomington,Illinois,61701,Central,Furniture,Tables,617.7000,...,-1.0,True,True,True,True,4,True,Extreme Discount,Negative Profit + Extreme Discount + Severe Ma...,56.705099


In [100]:
print("===== PROFITGUARD FINAL VALIDATION =====")

print("\n1. Duplicate rows:")
print(profit_leaks.duplicated().sum())

print("\n2. Missing values:")
print(profit_leaks.isnull().sum().sum())

print("\n3. Candidate count:")
print(len(profit_leaks))

print("\n4. Total candidate loss:")
print(profit_leaks["Profit"].sum())

print("\n5. Evidence Count:")
print(profit_leaks["Evidence_Count"].value_counts().sort_index())

print("\n6. Leak Type:")
print(profit_leaks["Leak Type"].value_counts())

print("\n7. Sub-Category:")
print(profit_leaks["Sub-Category"].value_counts())

print("\n8. Priority Score:")
print(
    profit_leaks["Subcategory_Priority"]
    .describe()
)

===== PROFITGUARD FINAL VALIDATION =====

1. Duplicate rows:
0

2. Missing values:
0

3. Candidate count:
328

4. Total candidate loss:
-106937.1006

5. Evidence Count:
Evidence_Count
2      7
3     17
4    304
Name: count, dtype: int64

6. Leak Type:
Leak Type
Extreme Discount      276
High Discount          43
Low Margin Loss         7
Severe Margin Loss      2
Name: count, dtype: int64

7. Sub-Category:
Sub-Category
Binders        139
Appliances      67
Tables          44
Bookcases       27
Machines        23
Furnishings     10
Phones           8
Storage          4
Chairs           3
Supplies         3
Name: count, dtype: int64

8. Priority Score:
count    328.000000
mean      53.822473
std       27.985413
min        8.203597
25%       18.253666
50%       56.705099
75%       81.838249
max       81.838249
Name: Subcategory_Priority, dtype: float64


In [102]:
profit_leaks.to_csv("profit_leaks.csv", index=False)

print("ProfitGuard dataset exported successfully!")
print("File: profit_leaks.csv")
print("Rows:", len(profit_leaks))
print("Columns:", len(profit_leaks.columns))

ProfitGuard dataset exported successfully!
File: profit_leaks.csv
Rows: 328
Columns: 25


In [103]:
profitguard_priority = leak_impact.reset_index()

profitguard_priority = profitguard_priority[
    [
        "Sub-Category",
        "Loss_Records",
        "Total_Loss",
        "Revenue",
        "Loss_Per_Record",
        "Loss_Margin",
        "Loss_Share",
        "Impact_Score",
        "Frequency_Score",
        "Severity_Score",
        "Priority_Score"
    ]
]

profitguard_priority.to_csv(
    "profitguard_priority.csv",
    index=False
)

print("Priority dataset exported!")
print("Rows:", len(profitguard_priority))
print("Columns:", len(profitguard_priority.columns))

Priority dataset exported!
Rows: 12
Columns: 11


In [104]:
region_leaks = (
    profit_leaks
    .groupby("Region")
    .agg(
        Leak_Records=("Profit", "count"),
        Revenue=("Sales", "sum"),
        Total_Loss=("Profit", "sum")
    )
    .sort_values("Total_Loss")
)

region_leaks["Loss_Share"] = (
    abs(region_leaks["Total_Loss"]) /
    abs(region_leaks["Total_Loss"]).sum()
) * 100

region_leaks

,Leak_Records,Revenue,Total_Loss,Loss_Share
Region,,,,
Central,212,53933.9048,-40546.1122,37.915851
East,54,66193.3220,-30869.1443,28.866637
South,27,62159.0640,-21184.0188,19.809793
West,35,16939.8520,-14337.8253,13.407718


### Actual Geographic Hotspots

In [105]:
state_leaks = (
    profit_leaks
    .groupby("State")
    .agg(
        Leak_Records=("Profit", "count"),
        Revenue=("Sales", "sum"),
        Total_Loss=("Profit", "sum")
    )
)

state_leaks["Loss_Share"] = (
    abs(state_leaks["Total_Loss"]) /
    abs(state_leaks["Total_Loss"]).sum()
) * 100

state_leaks = state_leaks.sort_values(
    "Total_Loss"
)

state_leaks.head(15)

,Leak_Records,Revenue,Total_Loss,Loss_Share
State,,,,
Texas,142,42592.9308,-26511.2703,24.791462
Ohio,18,20748.1160,-15571.9377,14.561773
Illinois,70,11340.9740,-14034.8419,13.124390
Pennsylvania,26,31932.9420,-12222.8624,11.429955
North Carolina,9,21527.3460,-9823.0363,9.185808
Colorado,14,7262.1140,-7760.7219,7.257277
Tennessee,8,10513.5280,-5825.6783,5.447762
Florida,10,30118.1900,-5535.3042,5.176224
Arizona,13,5920.4460,-4643.0441,4.341846


In [106]:
state_leaks.head(15)

,Leak_Records,Revenue,Total_Loss,Loss_Share
State,,,,
Texas,142,42592.9308,-26511.2703,24.791462
Ohio,18,20748.1160,-15571.9377,14.561773
Illinois,70,11340.9740,-14034.8419,13.124390
Pennsylvania,26,31932.9420,-12222.8624,11.429955
North Carolina,9,21527.3460,-9823.0363,9.185808
Colorado,14,7262.1140,-7760.7219,7.257277
Tennessee,8,10513.5280,-5825.6783,5.447762
Florida,10,30118.1900,-5535.3042,5.176224
Arizona,13,5920.4460,-4643.0441,4.341846


##  Now let's find the business root cause

<b>We've already established:</b>

Transaction → ML anomaly → candidate

Region → State → geographic concentration

Now we need to answer:

What products are actually driving these losses?

In [107]:
category_leaks = (
    profit_leaks
    .groupby(["Category", "Sub-Category"])
    .agg(
        Leak_Records=("Profit", "count"),
        Revenue=("Sales", "sum"),
        Total_Loss=("Profit", "sum")
    )
)

category_leaks["Loss_Share"] = (
    abs(category_leaks["Total_Loss"]) /
    abs(category_leaks["Total_Loss"]).sum()
) * 100

category_leaks = category_leaks.sort_values(
    "Total_Loss"
)

category_leaks

Leak_Records     Revenue  Total_Loss  Loss_Share
Category        Sub-Category                                                  
Office Supplies Binders                139  28753.5320 -32416.2865   30.313414
Technology      Machines                23  64989.8830 -28654.0276   26.795216
Furniture       Tables                  44  46093.5560 -19452.5599   18.190656
Office Supplies Appliances              67   3382.5340  -8629.6412    8.069829
Furniture       Bookcases               27  17933.7838  -8243.5945    7.708826
                Furnishings             10   2612.0800  -2983.5335    2.789989
Technology      Phones                   8  10795.8300  -2266.8778    2.119824
Office Supplies Supplies                 3   9926.3040  -2135.1666    1.996657
                Storage                  4   6128.9760  -1149.1830    1.074635
Furniture       Chairs                   3   8609.6640  -1006.2300    0.940955

In [109]:
## Discount Leaks

discount_leaks = (
    profit_leaks
    .groupby("Discount Band")
    .agg(
        Leak_Records=("Profit", "count"),
        Revenue=("Sales", "sum"),
        Total_Loss=("Profit", "sum")
    )
)

discount_leaks["Loss_Margin"] = (
    discount_leaks["Total_Loss"] /
    discount_leaks["Revenue"]
) * 100

discount_leaks["Loss_Share"] = (
    abs(discount_leaks["Total_Loss"]) /
    abs(discount_leaks["Total_Loss"]).sum()
) * 100

discount_leaks

,Leak_Records,Revenue,Total_Loss,Loss_Margin,Loss_Share
Discount Band,,,,,
10-20%,9,22066.6400,-3660.0196,-16.586212,3.422591
20-30%,4,10098.0880,-1303.9148,-12.912492,1.219329
30-40%,39,62078.7908,-17147.8612,-27.622737,16.035465
40%+,276,104982.6240,-84825.3050,-80.799376,79.322615


## Which products are losing money specifically under extreme discounting?

In [110]:
root_cause = (
    profit_leaks
    .groupby(["Sub-Category", "Discount Band"])
    .agg(
        Leak_Records=("Profit", "count"),
        Revenue=("Sales", "sum"),
        Total_Loss=("Profit", "sum")
    )
)

root_cause["Loss_Share"] = (
    abs(root_cause["Total_Loss"]) /
    abs(root_cause["Total_Loss"]).sum()
) * 100

root_cause = root_cause.sort_values(
    "Total_Loss"
)

root_cause.head(20)

,,Leak_Records,Revenue,Total_Loss,Loss_Share
Sub-Category,Discount Band,,,,
Binders,40%+,139,28753.5320,-32416.2865,30.313414
Machines,40%+,19,49810.0090,-26268.0461,24.564016
Tables,30-40%,22,29659.2300,-11152.1375,10.428689
Appliances,40%+,67,3382.5340,-8629.6412,8.069829
Tables,40%+,20,12942.3820,-7677.1656,7.179141
Bookcases,40%+,21,7482.0870,-6850.6321,6.406226
Furnishings,40%+,10,2612.0800,-2983.5335,2.789989
Machines,30-40%,4,15179.8740,-2385.9815,2.231201
Phones,30-40%,8,10795.8300,-2266.8778,2.119824


#  <b><u>Business Insight</u></b>

In [111]:
root_cause["Loss_Per_Record"] = (
    abs(root_cause["Total_Loss"]) /
    root_cause["Leak_Records"]
)

root_cause = root_cause.sort_values(
    "Loss_Per_Record",
    ascending=False
)

root_cause.head(15)

Leak_Records     Revenue  Total_Loss  Loss_Share  \
Sub-Category Discount Band                                                     
Machines     40%+                     19  49810.0090 -26268.0461   24.564016   
Supplies     10-20%                    3   9926.3040  -2135.1666    1.996657   
Machines     30-40%                    4  15179.8740  -2385.9815    2.231201   
Tables       30-40%                   22  29659.2300 -11152.1375   10.428689   
             40%+                     20  12942.3820  -7677.1656    7.179141   
Chairs       20-30%                    3   8609.6640  -1006.2300    0.940955   
Bookcases    40%+                     21   7482.0870  -6850.6321    6.406226   
Tables       10-20%                    1   2003.5200   -325.5720    0.304452   
Furnishings  40%+                     10   2612.0800  -2983.5335    2.789989   
Tables       20-30%                    1   1488.4240   -297.6848    0.278374   
Storage      10-20%                    4   6128.9760  -1149.1830    1.074635   
Phones       30-40%                    8  10795.8300  -2266.8778    2.119824   
Bookcases    30-40%                    5   6443.8568  -1342.8644    1.255752   
Binders      40%+                    139  28753.5320 -32416.2865   30.313414   
Appliances   40%+                     67   3382.5340  -8629.6412    8.069829   

                            Loss_Per_Record  
Sub-Category Discount Band                   
Machines     40%+               1382.528742  
Supplies     10-20%              711.722200  
Machines     30-40%              596.495375  
Tables       30-40%              506.915341  
             40%+                383.858280  
Chairs       20-30%              335.410000  
Bookcases    40%+                326.220576  
Tables       10-20%              325.572000  
Furnishings  40%+                298.353350  
Tables       20-30%              297.684800  
Storage      10-20%              287.295750  
Phones       30-40%              283.359725  
Bookcases    30-40%              268.572880  
Binders      40%+                233.210694  
Appliances   40%+                128.800615

## Final  Root-cause priority score

In [113]:
root_cause["Impact_Score"] = (
    root_cause["Loss_Share"] /
    root_cause["Loss_Share"].max()
) * 100

root_cause["Frequency_Score"] = (
    root_cause["Leak_Records"] /
    root_cause["Leak_Records"].max()
) * 100

root_cause["Severity_Score"] = (
    root_cause["Loss_Per_Record"] /
    root_cause["Loss_Per_Record"].max()
) * 100

root_cause["Priority_Score"] = (
    0.50 * root_cause["Impact_Score"] +
    0.30 * root_cause["Frequency_Score"] +
    0.20 * root_cause["Severity_Score"]
)

root_cause = root_cause.sort_values(
    "Priority_Score",
    ascending=False
)

root_cause.head(15)

Leak_Records     Revenue  Total_Loss  Loss_Share  \
Sub-Category Discount Band                                                     
Binders      40%+                    139  28753.5320 -32416.2865   30.313414   
Machines     40%+                     19  49810.0090 -26268.0461   24.564016   
Appliances   40%+                     67   3382.5340  -8629.6412    8.069829   
Tables       30-40%                   22  29659.2300 -11152.1375   10.428689   
             40%+                     20  12942.3820  -7677.1656    7.179141   
Bookcases    40%+                     21   7482.0870  -6850.6321    6.406226   
Supplies     10-20%                    3   9926.3040  -2135.1666    1.996657   
Machines     30-40%                    4  15179.8740  -2385.9815    2.231201   
Furnishings  40%+                     10   2612.0800  -2983.5335    2.789989   
Phones       30-40%                    8  10795.8300  -2266.8778    2.119824   
Chairs       20-30%                    3   8609.6640  -1006.2300    0.940955   
Bookcases    30-40%                    5   6443.8568  -1342.8644    1.255752   
Storage      10-20%                    4   6128.9760  -1149.1830    1.074635   
Tables       10-20%                    1   2003.5200   -325.5720    0.304452   
             20-30%                    1   1488.4240   -297.6848    0.278374   

                            Loss_Per_Record  Impact_Score  Frequency_Score  \
Sub-Category Discount Band                                                   
Binders      40%+                233.210694    100.000000       100.000000   
Machines     40%+               1382.528742     81.033483        13.669065   
Appliances   40%+                128.800615     26.621313        48.201439   
Tables       30-40%              506.915341     34.402884        15.827338   
             40%+                383.858280     23.683051        14.388489   
Bookcases    40%+                326.220576     21.133303        15.107914   
Supplies     10-20%              711.722200      6.586709         2.158273   
Machines     30-40%              596.495375      7.360441         2.877698   
Furnishings  40%+                298.353350      9.203810         7.194245   
Phones       30-40%              283.359725      6.993021         5.755396   
Chairs       20-30%              335.410000      3.104088         2.158273   
Bookcases    30-40%              268.572880      4.142561         3.597122   
Storage      10-20%              287.295750      3.545079         2.877698   
Tables       10-20%              325.572000      1.004347         0.719424   
             20-30%              297.684800      0.918319         0.719424   

                            Severity_Score  Priority_Score  
Sub-Category Discount Band                                  
Binders      40%+                16.868416       83.373683  
Machines     40%+               100.000000       64.617461  
Appliances   40%+                 9.316306       29.634350  
Tables       30-40%              36.665808       29.282805  
             40%+                27.764940       21.711060  
Bookcases    40%+                23.595934       19.818213  
Supplies     10-20%              51.479740       14.236785  
Machines     30-40%              43.145242       13.172578  
Furnishings  40%+                21.580264       11.076231  
Phones       30-40%              20.495757        9.322281  
Chairs       20-30%              24.260617        7.051649  
Bookcases    30-40%              19.426206        7.035658  
Storage      10-20%              20.780454        6.791940  
Tables       10-20%              23.549022        5.427805  
             20-30%              21.531907        4.981368